spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/ReleaseAnalyzer.java

```diff
@@ -101,45 +100,57 @@ private ReleaseAnalysisReport analyzeDependentPackagesOnly(
 			SpringCloudDeployerApplicationManifest matchingReplacingApplicationManifest = findMatching(
-					existingApplicationManifest.getApplicationName(),
-					replacingApplicationSpecList);
-			ReleaseDifference difference = compare(existingApplicationManifest,
-					matchingReplacingApplicationManifest);
-			if (!difference.areEqual()) {
-				logger.info(
-						"Dependent package difference found for " + existingApplicationManifest.getApplicationName());
-				appsToDelete.add(existingApplicationManifest.getApplicationName());
-				diffMessagesBuilder.append(difference.getDifferenceSummary());
-				diffMessagesBuilder.append("\n");
-			}
-		}
-		if (appsToDelete.size() != 0) {
-			ReleaseDifference releaseDifference = new ReleaseDifference(false, diffMessagesBuilder.toString());
-			return new ReleaseAnalysisReport(appsToDelete, releaseDifference, existingRelease, replacingRelease);
-		}
-		else {
-			ReleaseDifference releaseDifference = new ReleaseDifference(true);
-			return new ReleaseAnalysisReport(new ArrayList<>(), releaseDifference, existingRelease, replacingRelease);
+					applicationName, replacingApplicationSpecList);
+			ApplicationManifestDifference applicationManifestDifference = applicationManifestDifferenceFactory
+					.createApplicationManifestDifference(applicationName,
+							existingApplicationManifest,
+							matchingReplacingApplicationManifest);
+			applicationManifestDifferences.add(applicationManifestDifference);
 		}
+
+		return createReleaseAnalysisReport(existingRelease, replacingRelease, appsToDelete,
+				applicationManifestDifferences);
+
 	}
 
 	private ReleaseAnalysisReport analyzeTopLevelPackagesOnly(
 			List<? extends SpringCloudDeployerApplicationManifest> existingApplicationSpecList,
 			List<? extends SpringCloudDeployerApplicationManifest> replacingApplicationSpecList,
 			Release existingRelease, Release replacingRelease) {
-		ReleaseDifference difference = compare(existingApplicationSpecList.get(0),
-				replacingApplicationSpecList.get(0));
+
 		List<String> appsToDelete = new ArrayList<>();
-		if (!difference.areEqual()) {
-			logger.info(
-					"Differences detected, upgrading app " + existingApplicationSpecList.get(0).getApplicationName());
-			appsToDelete.add(existingApplicationSpecList.get(0).getApplicationName().trim());
-		}
-		return new ReleaseAnalysisReport(appsToDelete, difference, existingRelease, replacingRelease);
+		List<ApplicationManifestDifference> applicationManifestDifferences = new ArrayList<>();
+
+		ApplicationManifestDifference applicationManifestDifference = applicationManifestDifferenceFactory
+				.createApplicationManifestDifference(
+						existingApplicationSpecList.get(0).getApplicationName(),
+						existingApplicationSpecList.get(0),
+						replacingApplicationSpecList.get(0));
+		applicationManifestDifferences.add(applicationManifestDifference);
+
+		return createReleaseAnalysisReport(existingRelease, replacingRelease, appsToDelete,
```

Comment:
I think we can get rid of `appsToDelete` (also, this should have `appsToUpdate` I believe) local variable in these `analyze` methods as the appsToUpdate are now retrieved from the difference directly when creating the analysis report. Hence, this `appsToDelete` parameter can be removed as well.

Reply:
I will investigate.

Reply:
I see your point now after looking at the code.  I will move the creation of the list into the utility method `createReleaseAnalysisReport` (and rename it to appsToUpdate) but it will remain as a property of the `ReleaseAnalysisReport`.